File: R/LinearRegression.R
# +++++++++++++++++++++++++++++++++++++++++++++++++++++ #
# Implementing Linear Regression from Scratch           #
# Prediction                                            #                     
# Author: Jianhong Chen                                 #
# Date: 06.24.2021                                      #
#++++++++++++++++++++++++++++++++++++++++++++++++++++++ #

library(tidyverse)
library(matrixStats)


# ---------------------------- Model Modules -------------------------------

SyntheticData = function(w, num_examples){
  ## Generate y = Xw + b + noise 
  ## The intercept param, b,  is embedded into w as the first parameter!!!
  
  # NOTE set row-wise here to match with Python rowwise notation
  X = matrix(rnorm(num_examples * length(w), mean = 0, sd = 1),
             nrow = num_examples, ncol = length(w), byrow = TRUE)
  # set the first column into 1 for the intercept parameter
  X[, 1] = 1

  y = X %*% w  # matrix multiplication 
  y = y + rnorm(nrow(y), mean = 0, sd = 0.01) # add some noise
  
  ret = data.frame(y, X)
  
  return(ret)
  
}

LinReg = function(X, w){
  ## Linear Regression equation
  ## The intercept param, b,  is embedded into w as the first parameter!!!
  ret = X %*% w
  return(ret)
}

SquaredLoss = function(y_hat, y){
  ## Squared Loss of each iteration
  ret = (y_hat - y ) ^2 / 2
  return(ret)
}

SGD = function(X, y, w, lr, batch_size){
  ## minibatch stochastic gradient descent
  
  # Minibatch sampling:
  batch_ind = sample(nrow(X), batch_size, replace = FALSE)
  X_batch = X[batch_ind, ]
  y_batch = y[batch_ind, ]
  # Compute gradient for the batch sample:
  w_grad = matrix( (LinReg(X_batch, w) - y_batch), nrow = 1 ) # reshape for matrix operation 
  w_grad = matrix(w_grad %*% X_batch, ncol = 1) # reshape to col-vector

  # Now update w & b
  w = w - lr/batch_size * w_grad
  
  return(w)
  
}


# ---------------------------- Implementing the Model -------------------------------

## True parameters:
# note: the first parameter in w is the intercept
true_w = matrix(c(4.2, 2, -3.4), ncol = 1)
num_examples = 1000

dat = SyntheticData(true_w,  num_examples)
# VIZ synthetic data
dat %>% 
  ggplot(aes(x = X3, y = y)) +
  geom_point()

# initiate model parameters:
X = as.matrix(dat[, -1]) 
y = as.matrix(dat[, 1])

b = 0 
w = matrix(c(b, rnorm(2, mean = 0, sd = 0.01) ), ncol = 1)

batch_size = 125
lr = 0.5

num_epoch = 20
loss = SquaredLoss
net = LinReg

## -- Train the model:
for (epoch in 1:num_epoch ){
  # update parameters w:
  w = SGD(X, y, w, lr, batch_size)
  y_hat = net(X, w)
  l = loss(y_hat, y)
  
  print(str_c('epoch ', epoch, ' Loss ', mean(l) ))
  
}

print( str_c('True W: ', true_w, ' Optimized W: ', w))

# ------------------------------- Prediction -----------------------------------
y_pred = net(X, w)
# compute MAE and RMSE:
( mae = mean(abs(y - y_pred)) )
( rmse =  sqrt(mean((y-y_pred)^2)) )

data.frame(y, y_pred) %>%
  ggplot(aes(x = y, y = y_pred)) +
  geom_point() + 
  geom_abline(intercept = 0, slope = 1, lty = 2, color = 'red')


# ----------------------- Compare Built-in Function -------------------------------

mod_df = dat %>%
  dplyr::select(-X1) # remove the dummpy 1 column

mod = lm(y ~., data = mod_df)

print(mod$coefficients)
print(w)

File: Python/LinearRegression.py
'''
Implement Linear Regression from Scrach 

Author: Jianhong Chen
Date: 06.24.2021

'''

# %matplotlib inline


import random
from mxnet import autograd, np, npx
from d2l import mxnet as d2l

npx.set_np() # numpy set up

# -------------------------------------- Helper Functions ----------------------------


## -- Generating the Dataset:
def synthetic_data(w, b, num_examples): #@save
	""" Generate y = Xw + b + noise. """
	X = np.random.normal(0, 1, (num_examples, len(w)))
	y = np.dot(X, w) + b
	y += np.random.normal(0, 0.01, y.shape) # add noise to the output
	return X, y.reshape((-1,  1))

## -- Sampling batch of data
def data_iter(batch_size, features, labels):
	num_examples = len(features)
	indices = list(range(num_examples))
	# shuffle the original indices to make the sampling random:
	random.shuffle(indices)
	for i in range(0, num_examples, batch_size):
		batch_indices = np.array(indices[i:min(i+batch_size, num_examples)])
		# yield will return a generator type object
		yield features[batch_indices], labels[batch_indices]

## -- Defining the Model
def linreg(X, w, b): 
	""" Linear Regression Equation """
	return np.dot(X, w) + b

## -- Define the Loss Function
def squared_loss(y_hat, y):
	''' Squared loss. '''
	return (y_hat - y.reshape(y_hat.shape)) ** 2 / 2

## Define the Optimization Algorithm: SGD
def sgd(params, lr, batch_size):
	''' Minibatch stochastic gradient descent. '''
	for param in params:
		param[:] = param - lr/batch_size * param.grad 



## -- define true parameters 
true_w = np.array([2, -3.4])
true_b = 4.2

## -- Initiate data
features, labels = synthetic_data(true_w, true_b, 1000)
print('features:', features[0], '\nlabel:', labels[0])

# VIZ the synthetic data
d2l.set_figsize()
d2l.plt.scatter(features[:, (1)].asnumpy(), labels.asnumpy(), 1)
# d2l.plt.show()

batch_size = 10

for X, y in data_iter(batch_size, features, labels):
	print(X, '\n', y)
	break

## -- Initializaing Model Parameters:
w = np.random.normal(0, 0.01, (2, 1))
b = np.zeros(1)
w.attach_grad()
b.attach_grad()
# hyper-parameters:
lr = 0.03
num_epochs = 3
net = linreg
loss = squared_loss

# train the model
for epoch in range(num_epochs):
	for X, y in data_iter(batch_size, features, labels):
		with autograd.record():
			l = loss(net(X, w, b), y) # Minibatch loss in 'X' and 'y'
		l.backward()
		sgd([w, b], lr, batch_size) # update parameters using their gradient
	train_l = loss(net(features, w, b), labels)

	print(f'epoch {epoch + 1}, loss {float(train_l.mean()):f}')

print(f'error in estimating w: {true_w - w.reshape(true_w.shape)}')
print(f'error in estimating b: {true_b - b}')


